### PR TITLE
tui: navigation, scrolling, optimizations

### PR DESCRIPTION
**Pull request by vito**
[asciicast](https://asciinema.org/a/vdUEqRXX0K6Dw9zqLjkwlCxNc)

* ~~Now shows effects at their origin (`withExec`) rather than their unlazying point (`sync`).~~
  * Moved to separate PR: #7686 
* Now supports navigation, instead of being stuck with the bottom of the visible region
  * `hjkl` or arrow keys to move around
  * `enter` to "zoom in" to the selected span
  * `end` or `space` to follow the bottom of the screen
  * `home` to go to beginning
* Instead of having spans disappear after a second, we just collapse them when they finish.
* Now supports mouse scroll wheel again
  * I initially skipped this because I found it occasionally annoying for the TUI to steal mouse input. But you can just press Shift to bypass it.
* Now only renders the visible region, instead of rendering everything and cutting it off
  * This was a lot of work, but seems obviously worth it for large traces.
  * It's now possible to run our full integration suite without the TUI starting to lag. It used to get _real_ bad before.
* Now only renders logs when the logs need to be painted, rather than when the CLI receives them
* Tons of optimizations in the UI rendering loop to avoid repeated computations
* Avoid pushing cloud URL offscreen (it was also messing up the TUI a bit by printing directly to stdio)
* Now detects inifnite loop when printing simplified spans to avoid RAM usage explosion (cc @jedevc this should buy us some time)
* Now shows a keymap for better discoverability
* Now supports `+` and `-` keybinds increase and decrease verbosity at runtime
  * Want to know what's going on under `initialize`? Just hold `+`!

Here's a profile comparison for `dagger -m . call --source=.:default test important --race=true` after letting it run for ~9 minutes:

Before:

image

After:

image

Still spending a chunk of time in `WalkSpans`, but the total time in CPU is much shorter (23.58s -> 3.04s), and there's much less going on in there now (mostly just map usage).

### CODE REVIEW

File: dagql/idtui/frontend_plain.go
Comment (jedevc) [inline dagql/idtui/frontend_plain.go:248]: Mm, I have not actually found these logs of *particular* value - do we actually need any of these in `ExportSpans`/etc? They're quite noisy when enabled, and I'm not sure if there's a case where they're useful.

If we want to see the logs that the client receives, then it feels like maybe we should have that in the db?

File: go.mod
Comment (jedevc) [inline go.mod:8]: :shakes_fist_at_automated_tooling:

Love it, yeah, why not, makes total sense, this block definitely isn't separated for clarity :cry:

File: dagql/idtui/frontend_plain.go
Comment (vito) [reply dagql/idtui/frontend_plain.go:248]: As someone who spent the last month debugging OpenTelemetry draining issues I can attest to their value. :sweat_smile:

Could maybe move them to `ExtraDebug` level but every time we do this it only makes debugging OpenTelemetry more difficult for me. I would suggest just leaving these as-is until the coast feels clear.

> If we want to see the logs that the client receives, then it feels like maybe we should have that in the db?

Not sure what you mean here.

File: dagql/idtui/frontend_pretty.go
Comment (jedevc) [inline dagql/idtui/frontend_pretty.go:664]: Can we give some visual feedback when this happens? I was testing this, didn't realize it had opened, hammered the button, and went to my browser later to see 100s of cloud tabs open?

File: dagql/idtui/frontend_pretty.go
Comment (jedevc) [inline dagql/idtui/frontend_pretty.go:615]: Hm, something here is weird, it "feels" like CTRL+C now takes longer for the process to exit - there's a noticeable pause.

File: dagql/idtui/frontend_pretty.go
Comment (jedevc) [inline dagql/idtui/frontend_pretty.go:668]: Once we're zoomed, can we give some indication of the "context" we're in? Maybe like a breadcrumbs trail?

**Review by jedevc — COMMENTED** [review]
Left some little things, I like this though :tada: :tada:

File: go.mod
Comment (vito) [reply go.mod:8]: Yeah that's been annoying. Maybe if we move it below the other block it'll be safe? :thinking:

File: dagql/idtui/frontend.go
Comment (jedevc) [inline dagql/idtui/frontend.go:415]: :question: this doesn't seem to be working on `tty` progress anymore?

File: dagql/idtui/frontend_pretty.go
Comment (vito) [reply dagql/idtui/frontend_pretty.go:615]: It used to skip straight to an immediate interrupt because with the previous draining implementation it was too easy to end up in a scenario where you're stuck waiting on a "doomed" span (i.e. parent `kill -9`ed).

Now that that's fixed, I changed interrupt to not interrupt immediately, which gives telemetry a chance to drain, but takes a second or two. I can look into why. Maybe we have a 1 second sleep retry somewhere.

File: dagql/idtui/frontend_pretty.go
Comment (jedevc) [reply dagql/idtui/frontend_pretty.go:664]: Oooh, maybe we could also make the `Cloud` text a hyperlink :eyes:

File: dagql/idtui/frontend_plain.go
Comment (jedevc) [reply dagql/idtui/frontend_plain.go:242]: Yeah fair enough - worth keeping then!

> Not sure what you mean here.

I mean, we could move them into `ExportSpans` in `db.go`, that way they can be shared between the different frontends, instead. Just a suggestion, it may not work for logistical reasons :sweat_smile:

File: dagql/idtui/frontend.go
Comment (vito) [reply dagql/idtui/frontend.go:418]: The verbosity level defaults to 1, so this only kicks in when you press `-`

File: dagql/idtui/frontend_pretty.go
Comment (vito) [reply dagql/idtui/frontend_pretty.go]: Done! I think you might need to hold "Shift" for it to work though, not sure if worth it. Try it out in your terminal? This sort of thing varies a lot between implementations.

File: dagql/idtui/frontend_pretty.go
Comment (vito) [reply dagql/idtui/frontend_pretty.go:707]: Done! Now it'll pin your zoomed span to the top and nest the children, feels a lot more intuitive and avoids the feeling of having another UI element:

image

**Comment by vito** [comment]
@gerhard @jedevc Tossed this in the milestone since it seems ready and just needs approval, but feel free to skip it if you want to keep release scope tidy!

**Comment by jedevc** [comment]
Gonna bring this one out of the milestone - I think it's ready, but:
1. Ideally the next release to run will be v0.12.0 -  that fits nicely into here
2. Anything to improve the chances of giving a stable v0.11.8 :pray: :crossed_fingers: :cry:

File: dagql/idtui/frontend.go
Comment (jedevc) [inline dagql/idtui/frontend.go:412]: Are we intentionally not bringing this back before merging? Ideally I'd still like it, it made quite a few steps just never appear if they were quick enough.

File: dagql/idtui/frontend_pretty.go
Comment (jedevc) [inline dagql/idtui/frontend_pretty.go:328]: incredible nit: I got confused between `reset` and `quit` - `reset` goes to the top-level, and `quit` terminates everything.

I would probably expect `quit` to be at the end of the line as the last option, and maybe call `reset` something else? I can't think of a better name though, so take with a pinch of salt :salt:

File: dagql/idtui/frontend_pretty.go
Comment (jedevc) [reply dagql/idtui/frontend_pretty.go:707]: Yay :tada: feels like cloud now :tada:

File: dagql/idtui/frontend.go
Comment (vito) [reply dagql/idtui/frontend.go:412]: Didn't mean to leave this commented (it just needed access to more info post-refactor), but I do find the heuristic a little confusing and don't know how it should be updated for the new UI. Why do we want to show really fast spans beneath the primary span in particular, as opposed to the usual "really fast = boring/noise" logic? Does this affect the Plain UI more than the regular TUI? Does it affect the primary span in a way that's unique to the primary span and not other calls? Do we want this to apply to zoomed spans, too?

I'm trying to tease apart whether this is really a heuristic that belongs here or if it's working around a gap in other parts of the UI, like maybe when you select a span we can have a key to show its full recipe in a separate pane or something.

File: dagql/idtui/frontend_pretty.go
Comment (vito) [reply dagql/idtui/frontend_pretty.go]: Done!

* moved quit to the end
* reset -> unzoom, show only when zoomed

File: dagql/idtui/frontend.go
Comment (vito) [reply dagql/idtui/frontend.go]: Brought this back in the last commit to keep the PR unblocked but yeah, still find it a bit confusing. Found myself wondering if other things should also be revealed in this scenario (like internal steps) and couldn't intuit a rhyme or reason.

File: dagql/idtui/frontend.go
Comment (TomChv) [inline dagql/idtui/frontend.go:38]: I like this, will save a lot of logs, however will this still appears on cloud?

**Review by TomChv — COMMENTED** [review]
Just left a question, I'm going to try the PR manually then give an approval

**Comment by TomChv** [comment]
I gave it a try, this is absolutely amazing! Love it

I don't understand the `first` and `last` section in the context though, but I guess it's me missing something.

We should have a page on our doc with some records 
/cc @vikram-dagger 

That's absolutely amazing.

Uploading Screen Recording 2024-06-20 at 23.24.32.mov…

I also tried with failing pipelines, it works well, giving a nice output way clearer than before, this is an amazing improvement

**Review by TomChv — APPROVED** [review]

File: dagql/idtui/frontend.go
Comment (vito) [reply dagql/idtui/frontend.go:38]: Yeah this is only for the local TUI. I've experimented with a similar change for Cloud but haven't found a version I'm 100% happy with yet.

**Comment by vito** [comment]
> Is there some ways btw to tests that feature with UT? Like by checking the content of the SpanTree or asserts some values? That would be nice for later (not in this PR indeed)

Possibly yeah. I've also had good results with golden-style testing in Progrock, which let you cover a lot of behavior with fast tests + high confidence (you can visually see if it looks "right," catch regressions with ergonomic errors (you'll see the diff), and update them easily. ([example](https://github.com/vito/progrock/blob/main/tape_test.go) - though there's not much to see)

File: dagql/idtui/frontend.go
Comment (jedevc) [reply dagql/idtui/frontend.go]: Yeah, this is tricky, the heuristic definitely isn't perfect but there is a reason for it.

Suppose I have `dagger call do-thing process-it export --path=thing`. Ideally, my top-level spans should look *something* like:

- `MyModule`
- `MyModule.DoThing`
- `Thing.ProcessIt`
- `Directory.Export`

Each of these is tied directly to a CLI subcommand, so they should *always* show. The reason I originally noticed this was because the `export` was sometimes showing, sometimes not, and was essentially racy depending on the size of the file.

I think we should always show these user subcommand spans at the top level, which translated into the heuristic of always showing the top-level.

**Pull request merged**
